### PR TITLE
parity-util-mem: optimize MallocSizeOf for flat collections

### PR DESCRIPTION
**Pull request by ordian**
During investigation of slowness in `MemoryDB::size_of`, I've noticed that we iterate over collections when calculating size even if values have `size_of` of 0, e.g. for `HashMap<u32, u8>`.

To avoid iterations in this case, a new static method `fn size_of_is_zero()` is added to `MallocSizeOf`.
I've considered another approach with a marker trait, but it didn't work due to the lack of specialization.

I've noticed that the compiler is smart enough to optimize this for `Vec`'s already, but not for other collections: https://rust.godbolt.org/z/2jNo4c.

### CODE REVIEW

File: parity-util-mem/src/malloc_size.rs [hunk 1]
```diff
@@ -327,8 +349,10 @@ impl<T: MallocSizeOf> MallocSizeOf for [T] {
 impl<T: MallocSizeOf> MallocSizeOf for Vec<T> {
 	fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
 		let mut n = self.shallow_size_of(ops);
-		for elem in self.iter() {
-			n += elem.size_of(ops);
+		if !T::size_of_is_zero() {
+			for elem in self.iter() {
+				n += elem.size_of(ops);
+			}
```
Comment (NikVolf) [inline parity-util-mem/src/malloc_size.rs:355]:
```suggestion
	n = self.iter().fold(n, |acc, elem| acc + elem.size_of(ops))
```

**Comment by ordian** [comment]
> A possible variation would be to do :
fn constant_size() -> Option<usize>, which would allow other optimizations if the collection got a efficient method to return its size (usually).

I've thought about this as well, but unless `estimate-heapsize` feature is used, there is no way to know if a type actually allocates a fixed amount of bytes (even with `Box<[u8; 32]>` you never know it's 32 bytes with the allocator, that's why our tests have `>=` and not `==`).

**Comment by ordian** [comment]
> fn constant_size() -> Option<usize>, which would allow other optimizations if the collection got a efficient method to return its size (usually).

Implemented this because why not :)

**Comment by cheme** [comment]
just thought it could be an associated constant (locking trait object use case, but I am not sure if using mallocsizeof as trait object could be useful).

**Comment by ordian** [comment]
> just thought it could be an associated constant (locking trait object use case, but I am not sure if using mallocsizeof as trait object could be useful).

Is there any benefit of doing that? Both cases are optimized at compile at in release mode, so if there is no benefit, I'd prefer to keep as is and keep the ability to create `Box<dyn MallocSizeOf>` or `&dyn MallocSizeOf`.

**Comment by cheme** [comment]
Not that much, gives stronger guaranties, so we are sure there will be no bad implementation of the method.

**Comment by ordian** [comment]
> Not that much, gives stronger guaranties, so we are sure there will be no bad implementation of the method.

not sure I follow, could you elaborate how can no bad impl be guaranteed with the assoc const?

**Comment by cheme** [comment]
The associated constant strictly needs to be a constant.
On the other hand when implementing the function from a trait we do not need to return a constant value, someone can do anything (querying global mutable variable, calling ffi, rand...).
So using the associative constant is more restrictive.

**Comment by ordian** [comment]
> The associated constant strictly needs to be a constant.
> On the other hand when implementing the function from a trait we do not need to return a constant value, someone can do anything (querying global mutable variable, calling ffi, rand...).
> So using the associative constant is more restrictive.

I see, that's a contrived example though, an easier way to shoot oneself in the foot is to use `Some(usize::MAX)`. And I don't expect `constant_size` to be implemented at all, only via `malloc_size_is_0!` macro.

**Pull request merged**